firedrake/assemble.py

```diff
@@ -165,79 +185,165 @@ def create_assembly_callable(expr, tensor=None, bcs=None, form_compiler_paramete
+                             mat_type=mat_type,
+                             sub_mat_type=sub_mat_type,
+                             diagonal=diagonal,
+                             assembly_type="residual")
+
+
+def _get_assembly_rank(expr, diagonal):
+    """Return the assembly rank.
+
+    :arg expr: The expression being assembled.
+    :arg diagonal: If assembling a matrix is it diagonal?
+
+    :returns: The appropriate :class:`_AssemblyRank`.
+    """
+    rank = len(expr.arguments())
+    if diagonal:
+        assert rank == 2
+        return _AssemblyRank.VECTOR
+    if rank == 0:
+        return _AssemblyRank.SCALAR
+    if rank == 1:
+        return _AssemblyRank.VECTOR
+    if rank == 2:
+        return _AssemblyRank.MATRIX
+    raise AssertionError
+
+
+def _assemble_scalar(expr, scalar, bcs, opts):
+    """Assemble a scalar.
+
+    :arg expr: The expression being assembled.
+    :arg scalar: The scalar to write to (must be ``None``).
+    :arg bcs: Iterable of boundary conditions.
+    :arg opts: :class:`_AssemblyOpts` containing the assembly options.
+
+    :returns: The resulting :class:`float`.
+    """
+    if len(expr.arguments()) != 0:
+        raise ValueError("Can't assemble a 0-form with arguments")
+    if scalar:
+        raise ValueError("Can't assemble 0-form into existing tensor")
+
+    scalar = _make_scalar()
+    _assemble_expr(expr, scalar, bcs, opts, _AssemblyRank.SCALAR)
+    return scalar.data[0]
+
+
+def _assemble_vector(expr, vector, bcs, opts):
+    """Assemble a vector.
+
+    :arg expr: The expression being assembled.
+    :arg vector: The vector to write to (may be ``None``).
+    :arg bcs: Iterable of boundary conditions.
+    :arg opts: :class:`_AssemblyOpts` containing the assembly options.
+
+    :returns: The assembled :class:`firedrake.Function`.
+    """
+    if opts.diagonal:
+        test, trial = expr.arguments()
+        if test.function_space() != trial.function_space():
+            raise ValueError("Can only assemble diagonal of 2-form if functionspaces match")
+    else:
+        test, = expr.arguments()
+    if vector:
+        if test.function_space() != vector.function_space():
+            raise ValueError("Form's argument does not match provided result tensor")
+        vector.dat.zero()
+    else:
+        vector = _make_vector(test)
+
+    # Might have gotten here without EquationBC objects preprocessed.
+    if any(isinstance(bc, EquationBC) for bc in bcs):
+        bcs = tuple(bc.extract_form("F") for bc in bcs)
+
+    _assemble_expr(expr, vector, bcs, opts, _AssemblyRank.VECTOR)
+    return vector
+
+
+def _assemble_matrix(expr, matrix, bcs, opts):
+    """Assemble a matrix.
+
+    :arg expr: The expression being assembled.
+    :arg matrix: The matrix to write to (may be ``None``).
+    :arg bcs: Iterable of boundary conditions.
+    :arg opts: :class:`_AssemblyOpts` containing the assembly options.
+
+    :returns: The assembled :class:`.Matrix` or :class:`.ImplicitMatrix`.
```

Comment:
As for `_assemble_vector`, some comment on what the matrix _is_ would be welcome both here and in the `assemble` docstring. I.e. what DOFs the row and column spaces represent.

Reply:
I don't know the answer to this. @wence- @dham do you guys know?

Reply:
This is an internal function. There really isn't a need for pedagogical documentation here.

Reply:
What Reuben is asking for could live in the `assemble` docstring, but not sure if it currently does.

Reply:
I agree with @wence- , this should be in the `assemble` documentation if it isn't already. 

Nevertheless one should not assume that just because something isn't public that a developer will automatically understand what a function does. The documentation here can avoid defining what a matrix is but it would really help if it pointed out where to find out what it is.

Reply:
I've added a sentence to the docstring for `assemble` and pointed to it from this function.